crates/kubelet/src/logs.rs

```diff
@@ -0,0 +1,179 @@
+use anyhow::bail;
+use log::{debug, error};
+use tokio::io::{AsyncBufReadExt, AsyncRead};
+
+/// Sender for streaming logs to client.
+pub struct LogSender {
+    sender: Option<hyper::body::Sender>,
+    tail: Option<usize>,
+    follow: bool,
+}
+
+/// Possible errors sending log data.
+#[derive(Debug)]
+pub enum LogSendError {
+    /// Client has disconnected.
+    ChannelClosed,
+    /// An unexpected error occured.
+    Abnormal(anyhow::Error),
+}
+
+impl From<std::io::Error> for LogSendError {
+    fn from(error: std::io::Error) -> Self {
+        LogSendError::Abnormal(anyhow::Error::new(error))
+    }
+}
+
+impl std::fmt::Display for LogSendError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            LogSendError::ChannelClosed => write!(f, "ChannelClosed"),
+            LogSendError::Abnormal(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+impl std::error::Error for LogSendError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            LogSendError::ChannelClosed => None,
+            LogSendError::Abnormal(e) => Some(e.root_cause()),
+        }
+    }
+}
+
+impl LogSender {
+    /// Create new `LogSender` from `hyper::body::Sender`.
+    pub fn new(sender: hyper::body::Sender, tail: Option<usize>, follow: bool) -> Self {
+        LogSender {
+            sender: Some(sender),
+            tail,
+            follow,
+        }
+    }
+
+    /// The tail flag indicated by the request if present.
+    pub fn tail(&self) -> Option<usize> {
+        self.tail.clone()
+    }
+
+    /// The follow flag indicated by the request, or `false` if absent.
+    pub fn follow(&self) -> bool {
+        self.follow
+    }
+
+    /// Async send some data to a client.
+    pub async fn send(&mut self, data: String) -> Result<(), LogSendError> {
+        let b: hyper::body::Bytes = data.into();
+        match self.sender {
+            Some(ref mut sender) => sender.send_data(b).await.map_err(|e| {
+                if e.is_closed() {
+                    debug!("channel closed.");
+                    LogSendError::ChannelClosed
+                } else {
+                    error!("channel error: {}", e);
+                    LogSendError::Abnormal(anyhow::Error::new(e))
+                }
+            }),
+            None => Err(LogSendError::ChannelClosed),
+        }
+    }
+
+    /// Gracefully close the channel.
+    pub fn close(&mut self) {
```

Comment:
Nit: Should this just be implemented as `Drop`?

Reply:
This should be the default implementation of drop. Do we need this function at all?

Reply:
Done.